### PR TITLE
git: Add option to use diff-so-fancy as diff tool

### PR DESCRIPTION
**Pull request by arjan-s**
### Description
Add the option to use [diff-so-fancy](https://github.com/so-fancy/diff-so-fancy) as diff tool in git. This makes the shown diffs human readable (and pretty) instead of machine readable.
This PR was inspired by https://github.com/nix-community/home-manager/pull/2850

### Checklist

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all`.

- [ ] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/docs/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

  - [ ] Added myself and the module files to `.github/CODEOWNERS`.

### CODE REVIEW

File: modules/programs/git.nix [hunk 1]
```diff
@@ -296,16 +296,71 @@ in {
...
   };
 
   config = mkIf cfg.enable (mkMerge [
     {
       home.packages = [ cfg.package ];
       assertions = [{
-        assertion = !(cfg.delta.enable && cfg.difftastic.enable);
+        assertion = !(cfg.delta.enable && cfg.difftastic.enable)
+          && !(cfg.delta.enable && cfg.diff-so-fancy.enable)
+          && !(cfg.difftastic.enable && cfg.diff-so-fancy.enable);
```
Comment (ambroisie) [inline modules/programs/git.nix:361]: You might just want to put them in a list, and count that at most one is enabled.

Comment (arjan-s) [reply modules/programs/git.nix:361]: @ambroisie Thanks for your review! I actually wanted to implement it that way, but couldn't figure out how to do it in Nix... Can you help with that?

Comment (olmokramer) [reply modules/programs/git.nix:361]: @arjan-s You could do this for example

```
assertion = let
  enabled = [
    cfg.delta.enable
    cfg.diff-so-fancy.enable
    cfg.difftastic.enable
  ];
in lib.length (lib.filter lib.id enabled) == 1;
```

**Review by ambroisie — APPROVED** [review]

File: modules/programs/git.nix [hunk 2]
```diff
@@ -296,16 +296,75 @@ in {
...
+          description = ''
+            Simplify git header chunks to a more human readable format.
+          '';
+        };
+        stripLeadingSymbols = mkOption {
+          type = types.bool;
+          default = true;
+          example = false;
+          description = ''
+            Should the pesky + or - at line-start be removed.
```
Comment (sumnerevans) [inline modules/programs/git.nix:328]:
```suggestion
            Should the pesky <literal>+</literal> or <literal>-</literal> at line-start be removed?
```

File: modules/programs/git.nix [hunk 3]
```diff
@@ -296,16 +296,75 @@ in {
...
+            Enable the <command>diff-so-fancy</command> diff colorizer.
+            See <link xlink:href="https://github.com/so-fancy/diff-so-fancy" />.
+          '';
+        };
+        markEmptyLines = mkOption {
+          type = types.bool;
+          default = true;
+          example = false;
+          description = ''
+            Should the first block of an empty line be colored.
```
Comment (sumnerevans) [inline modules/programs/git.nix:312]:
```suggestion
            Should the first block of an empty line be colored?
```

**Review by sumnerevans — CHANGES_REQUESTED** [review]

File: modules/programs/git.nix (hunk 3 above)
Comment (rycee) [reply modules/programs/git.nix]: I think it feels a bit odd for the documentation to ask the user questions. I would suggest to instead use the "Whether …" kind of phrasing. E.g.

> Whether the first block of an empty line should be colored.

File: modules/programs/git.nix [hunk 4]
```diff
@@ -296,16 +296,75 @@ in {
...
       };
+
+      diff-so-fancy = {
+        enable = mkEnableOption "" // {
+          description = ''
+            Enable the <command>diff-so-fancy</command> diff colorizer.
+            See <link xlink:href="https://github.com/so-fancy/diff-so-fancy" />.
+          '';
+        };
+        markEmptyLines = mkOption {
```
Comment (rycee) [inline modules/programs/git.nix:307]: Please make sure to separate option definitions by an empty line to make the module more readable. That is,
```suggestion

        markEmptyLines = mkOption {
```

File: modules/programs/git.nix [hunk 5]
```diff
@@ -296,16 +296,75 @@ in {
...
       home.packages = [ cfg.package ];
       assertions = [{
-        assertion = !(cfg.delta.enable && cfg.difftastic.enable);
+        assertion = let
+          enabled = [
+            cfg.delta.enable
+            cfg.diff-so-fancy.enable
+            cfg.difftastic.enable
+          ];
+        in lib.length (lib.filter lib.id enabled) <= 1;
```
Comment (rycee) [inline modules/programs/git.nix:365]:
```suggestion
        in count id enabled <= 1;
```

File: modules/programs/git.nix (hunk 3 above)
Comment (arjan-s) [reply modules/programs/git.nix]: > I think it feels a bit odd for the documentation to ask the user questions. I would suggest to instead use the "Whether …" kind of phrasing. E.g.
> 
> > Whether the first block of an empty line should be colored.

Yeah, I agree. I first opted to use the option descriptions from the original readme, but this seems nicer.

**Pull request merged**